### PR TITLE
feat: add pages `/items/my` and `/items/borrowed`

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes https://github.com/hpi-swt2/bookkeeper-portal-red/issues/79
Closes https://github.com/hpi-swt2/bookkeeper-portal-red/issues/78

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 Rails.application.routes.draw do
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
+  get "/items/my", to: "items#my_items"
+  get "/items/borrowed", to: "items#borrowed_items"
```
Comment (Skn0tt) [inline config/routes.rb:5]: It feels a bit weird to extend `/items` like this, given it's declared as a resource. @luap42 what would be the Rails-idiomatic way of doing this?

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -19,6 +19,18 @@ def new
...
 
+  # GET /items/borrowed
+  def borrowed_items
+    @items = current_user.lendings.where(completed_at: nil).map(&:item)
+    render :borrowed_items
+  end
+
+  # GET /items/my
+  def my_items
+    @items = Item.all # TODO: how can we find out which items are owned by the current user?
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply app/controllers/items_controller.rb:34]: to the team that takes this over, here's the response that @Shieka received:

> Ja stimmt, dieses "Feature" ist jetzt in die Permission "can_manage" gewandert. Wenn eine Gruppe ein neues Item anlegt, sollte diese Permission automatisch angelegt werden."
 unterm Strich "gehören" Items jetzt allen, die die permission "can_manage" haben

File: spec/features/items/my_items_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  let(:user) { FactoryBot.create(:user, password: password) }
+  let(:group) { FactoryBot.create(:group) }
+
+  # for some reason, this does not work and fails with the following exception
+  # once i try to look at the membership in the debugger:
+  # "error: undefined method `status=' for Membership"
+
+  let(:membership) { FactoryBot.create(:membership, user: user, group: group, role: 1) }
+  let(:item1) { FactoryBot.create(:item, name: 'Item 1') }
+  let(:item2) { FactoryBot.create(:item, name: 'Item 2') }
```
Comment (JannisHajda) [inline spec/features/items/my_items_spec.rb:14]: The feature itself should be implemented with the latest commit (16223c00be028a792e0521b1049931184a9ce253), but I can't seem to get the tests to be working. For some reason, creating a membership results in an exception, as soon as I try to take a look at it in the debugger. 

Would be really great, if one of you guys could take a look at it! @phkeese @yannik-dittmar

**Comment by JannisHajda** [comment]
Okay, got the tests working. Pull request is now ready for review! @luap42

File: app/controllers/items_controller.rb [hunk 4]
```diff
@@ -23,6 +23,28 @@ def new
   def edit
   end
 
+  # GET /items/borrowed
+  # lists all items borrowed by the current user
+  def borrowed_items
```
Comment (luap42) [inline app/controllers/items_controller.rb:28]: - style: IMO the "items" in this route name is redundant because the controller already implies that we are talking about items; suggestion to rename to, for example, "borrowed_by_me"

File: app/controllers/items_controller.rb [hunk 5]
```diff
@@ -23,6 +23,28 @@ def new
...
+    @items = current_user.lendings.where(completed_at: nil).map(&:item)
+    render :borrowed_items
+  end
+
+  # GET /items/my/borrowed
+  # lists all items of the current user which are currently borrowed
+  def my_borrowed_items
```
Comment (luap42) [inline app/controllers/items_controller.rb:35]: - style: (same as above), suggestion to rename to, for example, "mine_borrowed"

File: app/controllers/items_controller.rb [hunk 6]
```diff
@@ -23,6 +23,28 @@ def new
...
+    # return items which are currently not lendable
+    @items = current_user.items.reject(&:lendable?)
```
Comment (luap42) [inline app/controllers/items_controller.rb:37]: - optimization: AFAICT `#reject` works app-side; this is probably fine for now, but if there might be huge owners with hundreds or thousands of own items, this check should probably be deferred to the database.

- style: this code returns all items that are not lendable; right now, these are only borrowed items, but for example in the future we might add something like locked/damaged state, which also precludes lendability without implying borrowedness, so it is probably preferable to have a separate `#borrowed?` method and then make `#lendable?` to `!borrowed?`.

- structure: I also think that the lendable? check is something, where Rails scopes (https://guides.rubyonrails.org/active_record_querying.html#scopes) might be a good fit, please check out.

File: app/controllers/items_controller.rb [hunk 7]
```diff
@@ -23,6 +23,28 @@ def new
   def edit
   end
 
+  # GET /items/borrowed
+  # lists all items borrowed by the current user
+  def borrowed_items
+    @items = current_user.lendings.where(completed_at: nil).map(&:item)
+    render :borrowed_items
```
Comment (luap42) [inline app/controllers/items_controller.rb:30]: - redundancy: Rails should automatically render the `borrowed_items.html.erb` upon GET request to this route.

File: app/controllers/items_controller.rb [hunk 8]
```diff
@@ -23,6 +23,28 @@ def new
...
+    # return items which are currently not lendable
+    @items = current_user.items.reject(&:lendable?)
+    render :borrowed_items
+  end
+
+  # GET /items/my
+  # lists all items of the current user
+  def my_items
+    @items = current_user.items
+    render :my_items
```
Comment (luap42) [inline app/controllers/items_controller.rb:45]: - redundancy (see above)

File: app/models/user.rb [hunk 9]
```diff
@@ -27,6 +27,11 @@ def lending_rights?(item)
     false
   end
 
+  def items
+    # get all items where the user is a manager of any group
+    Item.joins(:manager_groups).where(groups: { id: groups })
```
Comment (luap42) [inline app/models/user.rb:32]: - structure (optional): if possible, it might be preferable to use has_many relations for clarity

File: app/views/items/_item_grid.html.erb [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<div id="items">
+  <div class="container">
+      <div class="row ">
+        <% @items.each do |item| %>
```
Comment (luap42) [inline app/views/items/_item_grid.html.erb:4]: - style: the item grid should not be dependent on a proper global variable (`@items`) to be set; instead you can pass arguments through the `render` helper, which should be used to include a common API; this makes it possible to use the item grid multiple times on one page or in different contexts and is therefore IMO preferable.

File: app/views/items/_item_grid.html.erb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+          <div class="col-sm-3 mb-4">
+          <div class="card">
+            <img class="card-img-top" src="https://via.placeholder.com/400x300" alt="Card image cap">
+            <div class="card-body text-center">
+              <h5 class="card-title text-truncate"><%= item.name %></h5>
+              <p class="card-text text-truncate">von: Alan Must</p>
```
Comment (luap42) [inline app/views/items/_item_grid.html.erb:10]: - question: what is the purpose/expected life time of these "example data"? If there are no data yet, there should be no data displayed.

File: app/views/items/_item_grid.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+              <p class="card-text text-truncate text-success">verf&uuml;gbar</p>
```
Comment (luap42) [inline app/views/items/_item_grid.html.erb:11]: - incomplete: same goes here, but in this case we do have the data with the `#lendable?` operator; to fix it should be just a simple ternary like `<?= item.lendable? ? 'verfügbar' : 'verliehen' ?>` (might require consequential changes due to `#borrowed?` mentioned in the review comment above)

File: app/views/items/_item_grid.html.erb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+              <%= link_to "Show item", item, class: "card-link" %>
+              <%= link_to "Edit item", edit_item_path(item), class: "card-link" %>
```
Comment (luap42) [inline app/views/items/_item_grid.html.erb:13]: - i18n: please use consistently either English or German; I recognize that this might not be the case in other views, but do not introduce new discrepancies in views that are being created/modified in this MR.

File: app/views/items/borrowed_items.html.erb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<p style="color: green"><%= notice %></p>
+
+<h1>Borrowed Items</h1>
+
+<%= render "item_grid" %>
```
Comment (luap42) [inline app/views/items/borrowed_items.html.erb:5]: - style: include variable; see above

File: app/views/items/borrowed_items.html.erb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<p style="color: green"><%= notice %></p>
```
Comment (luap42) [inline app/views/items/borrowed_items.html.erb:1]: - question: What is the purpose of this line? If this is intended for status notifications, such as confirmation of login/logout or of requests, please put it into the layout, because it is then likely to be the same on every page.

File: app/views/items/index.html.erb [hunk 16]
```diff
@@ -1,24 +1,5 @@
...
-              <%= link_to "Show item", item, class: "card-link" %>
-              <%= link_to "Edit item", edit_item_path(item), class: "card-link" %>
-            </div>
-          </div>
-          </div>
-        <% end %>
-      </div>
-    </div>
-</div>
+<%= render "item_grid" %>
```
Comment (luap42) [inline app/views/items/index.html.erb:3]: - style: include variable; see above

File: app/views/items/my_items.html.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+<p style="color: green"><%= notice %></p>
+
+<h1>My Items</h1>
+
+<%= render "item_grid" %>
```
Comment (luap42) [inline app/views/items/my_items.html.erb:5]: - style: include variable; see above

File: app/views/items/my_items.html.erb [hunk 18]
```diff
@@ -0,0 +1,5 @@
+<p style="color: green"><%= notice %></p>
```
Comment (luap42) [inline app/views/items/my_items.html.erb:1]: - question: (see above)

File: app/views/items/my_items.html.erb [hunk 19]
```diff
@@ -0,0 +1,5 @@
...
+
+<h1>My Items</h1>
```
Comment (luap42) [inline app/views/items/my_items.html.erb:3]: - i18n: On language, see above

File: app/views/items/borrowed_items.html.erb [hunk 20]
```diff
@@ -0,0 +1,5 @@
...
+
+<h1>Borrowed Items</h1>
```
Comment (luap42) [inline app/views/items/borrowed_items.html.erb:3]: - i18n: On language, see above

File: config/routes.rb [hunk 21]
```diff
@@ -3,6 +3,11 @@
 
   resources :items do
     patch :update_lending
+    collection do
+      get "/my", to: "items#my_items"
+      get "/my/borrowed", to: "items#my_borrowed_items"
+      get "/borrowed", to: "items#borrowed_items"
```
Comment (luap42) [inline config/routes.rb:9]: - structure: please add aliases (`as: :some_label`) for these routes, so that they can be addressed validly within the templates

File: db/schema.rb [hunk 22]
```diff
@@ -77,10 +77,10 @@
     t.datetime "remember_created_at"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
-    t.string "full_name", default: "", null: false
-    t.string "description", default: "", null: false
     t.string "provider", limit: 50, default: "", null: false
     t.string "uid", limit: 50, default: "", null: false
+    t.string "full_name", default: "", null: false
+    t.string "description", default: "", null: false
```
Comment (luap42) [inline db/schema.rb:83]: - question: Why this change?
